### DISCUSSION TITLE
GPU?

### DISCUSSION THREAD

**Discussion by M-O-Marmalade** [4ms/stm32mp1-baremetal#32] [Q&A] [answered]
Hello! First off, thank you for providing all the work you've accomplished here in this repository! Really amazing stuff!

I'm wondering if it's possible to access the GPU on the STM32MP157D while using your bare metal framework, using [Mesa](https://www.mesa3d.org/)?

Mesa is able to access the STM32MP157D's GPU on kernels/OSs such as Linux and the BSDs. From looking at the Mesa [source code](https://archive.mesa3d.org/?C=M;O=D), it doesn't look like it relies on any external OS components, but it says [here](https://docs.mesa3d.org/index.html): 

> "Mesa ties into several other open-source projects: the [Direct Rendering Infrastructure](https://dri.freedesktop.org/), [X.org](https://x.org/), and [Wayland](https://wayland.freedesktop.org/) to provide OpenGL support on Linux, FreeBSD, and other operating systems." 

This has me a little confused as to whether Mesa _requires_ an OS, or just _provides additional support_ for OSs. I wonder if it would be possible to just statically link Mesa's source into my application and build an image that can be booted and run via your bare metal framework?

Anyways, thank you for any knowledge you might be able to impart on the subject!

**Comment by danngreen** [comment, marked as answer]
Well, I know very little about how Mesa works.

But I would think you would need a hardware driver for it to interface with the GPU. That might be satisfied with this:
https://github.com/STMicroelectronics/gcnano-binaries/tree/gcnano-6.4.15-binaries/gcnano-driver-stm32mp/hal

I don't know what the interface between Mesa and the gcnano driver would be, though.

> From looking at the Mesa source code, it doesn't look like it relies on any external OS components

That would be nice! I guess the best way to find out would be to compile it with arm-none-eabi toolchain and see what happens. Did you see any docs or reports if compiling it as a static library can work (as opposed to shared lib)?
It's actually not impossible to load shared objects, we did it for the MP1 on the [MetaModule project](https://github.com/4ms/metamodule) but it's complicated and requires a custom built standard library (and I still haven't gotten the entire c++ std lib ported).
